core/preconditioner/schwarz.cpp

```diff
@@ -0,0 +1,208 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/preconditioner/schwarz.hpp>
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/precision_dispatch.hpp>
+#include <ginkgo/core/base/temporary_conversion.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/utils.hpp"
+
+
+namespace gko {
+namespace preconditioner {
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    precision_dispatch_real_complex<ValueType>(
+        [this](auto dense_b, auto dense_x) {
+            this->apply_dense_impl(dense_b, dense_x);
+        },
+        b, x);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Schwarz<ValueType, IndexType>::apply_dense_impl(
+    const matrix::Dense<ValueType>* dense_b,
+    matrix::Dense<ValueType>* dense_x) const
+{
+    using Vector = matrix::Dense<ValueType>;
+    auto exec = this->get_executor();
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+    size_type offset = 0;
+
+    // FIXME: Sequential solution of subdomains.
+    for (size_type i = 0; i < this->num_subdomains_; ++i) {
+        size_type l_num_rows = this->subdomain_matrices_[i]->get_size()[0];
+        auto rspan = gko::span(offset, offset + l_num_rows);
+        const auto b_view = dense_b->create_submatrix(
+            rspan, gko::span(0, dense_b->get_size()[1]));
+        auto x_view = dense_x->create_submatrix(
+            rspan, gko::span(0, dense_x->get_size()[1]));
+        this->subdomain_solvers_[i]->apply(b_view.get(), x_view.get());
+        offset += l_num_rows;
+    }
+    if (parameters_.coarse_solver.uses_coarse_solver_) {
+        auto r = Vector::create_with_config_of(dense_b);
+        r->copy_from(dense_b);
+        auto restrict_op =
+            parameters_.coarse_solver.coarse_op_->get_restrict_op();
```

Comment:
Since this is pretty long, you could also store a pointer to the `coarse_op_` in this class?